observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingChannelInterceptor.java

```diff
@@ -16,33 +16,173 @@
 
 package io.grpc.observability.interceptors;
 
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Durations;
 import io.grpc.CallOptions;
 import io.grpc.Channel;
 import io.grpc.ClientCall;
 import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Deadline;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
 import io.grpc.Internal;
+import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.interceptors.GcpLogHelper.GcpLogSinkWriter;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventLogger;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
 
-/** A logging interceptor for {@code LoggingChannelProvider}. */
+/**
+ * A logging interceptor for {@code LoggingChannelProvider}.
+ */
 @Internal
 public final class InternalLoggingChannelInterceptor implements ClientInterceptor {
+  private final GcpLogSinkWriter writer;
 
   public interface Factory {
     ClientInterceptor create();
   }
 
   public static class FactoryImpl implements Factory {
+    Sink sink;
+
+    public FactoryImpl(Sink sink) {
+      this.sink = sink;
+    }
 
     @Override
     public ClientInterceptor create() {
-      return new InternalLoggingChannelInterceptor();
+      return new InternalLoggingChannelInterceptor(sink);
     }
   }
 
+  private InternalLoggingChannelInterceptor(Sink sink) {
+    this.writer = new GcpLogSinkWriter(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
+  }
+
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
       CallOptions callOptions, Channel next) {
-    // TODO(dnvindhya) implement the interceptor
-    return null;
+
+    final AtomicLong seq = new AtomicLong(1);
+    final String rpcId = UUID.randomUUID().toString();
+    final String authority = next.authority();
+    final String serviceName = method.getServiceName();
+    final String methodName = method.getBareMethodName();
+    final Deadline deadline = GcpLogHelper.min(callOptions.getDeadline(),
```

Comment:
Can you add a comment about using `min` on `callOptions.getDeadline()` and `Context.current().getDeadline()` ?

Reply:
Added